cache.go

```diff
@@ -0,0 +1,79 @@
+package rsmt2d
+
+import (
+	"reflect"
+	"sync"
+)
+
+var (
+	defaultDoubleCacheCapacity = 1024 * 1024
+)
+
+type doubleCacheOptions struct {
+	capacity int
+}
+
+func defaultDoubleCacheOptions() doubleCacheOptions {
+	return doubleCacheOptions{
+		capacity: defaultDoubleCacheCapacity,
+	}
+}
+
+func (opts doubleCacheOptions) withCapacity(capacity int) doubleCacheOptions {
+	opts.capacity = capacity / 2
+	return opts
+}
+
+type doubleCache[T any] struct {
+	opts           doubleCacheOptions
+	cacheMu        *sync.Mutex
+	cacheFrontSize int
+	cacheFront     map[int]T
+	cacheBack      map[int]T
```

Comment:
Maybe a dumb question but I do not fully understand why a double cache makes sense here. The access will be fast as internally we are accessing a map anyways (the complexity will asymptotically be the same in the front as well as in the back). Wouldn't single cache be as good as this and would be much simpler to follow? 
I almost feel a [sync.Map](https://pkg.go.dev/sync#Map) would have served the same purpose. Or maybe a tiny wrapper to make it bounded. But maybe I am missing something obvious here.

Reply:
I've used a sync.Map here: https://github.com/celestiaorg/rsmt2d/pull/136

Does the cache need to be strictly bounded? I think that is not really necessary as a rather small number of different sizes will occur over time and for Celestia specifically it will be a bunch of powers-of-two (cc @musalbas)

Reply:
The motivation behind the double cache was to keep it bounded and keep only encoders that are accessed frequently relevant

Reply:
🤔  I think in practice less encoders than the default bound will be used anyways:
https://github.com/celestiaorg/rsmt2d/blob/e454f365b8e029fb147a656e2145dc916c1ba659/cache.go#L9

That said, I didn’t fully think this through but we should. Also what is the mem-footprint of one encoder.
Also I think the capacity isn’t really adjustable from the outside anymore. So this is more like a constant than an option in any case btw.